### ISSUE TITLE
Unresolved import for platform specific modules

### ISSUE THREAD

**Issue by GrayJack** [rust-analyzer/rust-analyzer#6038]
I tried to search if it was reported but didn't found anything.

When importing the platform specific modules in `std::os` rust-analyzer gives me the following error:
`unresolved import`. Also those modules don't show up for auto-complete.

When the module is used outside `use` statement, there is no error, but the semantic token are of the type `unresolvedReference`.

### RA version
Latest master branch

### Example
```rust
use std::os::unix;
```

**Comment by jonas-schievink** [comment]
We seem to have trouble finding/expanding the `cfg_if!` invocation

**Comment by jonas-schievink** [comment]
Oh, it's because we don't load the sysroot correctly. We just add a bunch of hard-coded path dependencies to libstd without looking at its `Cargo.toml`.

**Comment by jfcha** [comment]
I'm also seeing this issue for 

`use std::os::windows::ffi::{OsStrExt, OsStringExt};`

in the 2020-09-21 release.

**Comment by ttzztztz** [comment]
Same issue.

**Comment by iulianR** [comment]
Hello, it's a bit spammy in VSCode (using Code Lens extension) because it looks like a hard error, making you think the code won't compile:
![image](https://user-images.githubusercontent.com/6775259/94371759-89988000-0101-11eb-9f7c-042c275f57e3.png)
![image](https://user-images.githubusercontent.com/6775259/94371760-90bf8e00-0101-11eb-9b3e-ef579afa1a87.png)
![image](https://user-images.githubusercontent.com/6775259/94371697-0ecf6500-0101-11eb-9e87-2209f1821016.png)

It is possible to change it somehow it so it shows in a similar way to an `{unknown}` error (when `ra` is unable to determine type for example)? I think that was the previous behaviour before the 21-09-2020 release.
Thanks!

**Comment by lnicola** [comment]
@iulianR you can disable that warning in the preferences (the experimental diagnostics key).

**Comment by leshow** [comment]
I am running into this too, if I use the above disable though, won't that disable all unresolved import errors?

That doesn't seem to be desirable. There should be some way to do a platform specific import?

**Comment by jonas-schievink** [comment]
No, it happens because

> we don't load the sysroot correctly. We just add a bunch of hard-coded path dependencies to libstd without looking at its `Cargo.toml`.

You can only disable the unresolved import diagnostic entirely (although check-on-save will still display rustc's unresolved import messages, regardless of how r-a diagnostics are configured)

**Comment by thomcc** [comment]
This is (unsurprisingly) also broken for `{core,std}::arch::*`. I'm certain this used to work too... :(

**Comment by flodiebold** [comment]
It "used to work" in the sense that this diagnostic didn't show up, because the diagnostic wasn't implemented yet. You can make it not show up again with the setting mentioned above; you'll still get unresolved import errors from rustc.

**Comment by jonas-schievink** [comment]
The diagnostic is already marked as experimental fwiw, so unless you opt into the experimental ones this shouldn't show up.

**Comment by thomcc** [comment]
> It "used to work" in the sense that this diagnostic didn't show up

I guess I'll take your word for it that it never worked, but I was pretty sure I had autocomplete for some parts of core::arch in the past. Huh.

**Comment by jonas-schievink** [comment]
Oh, turns out the experimental ones are on by default

**Comment by cameron-rowe** [comment]
> 
> 
> Oh, turns out the experimental ones are on by default

Disabling that caused the error to go away for me. This works as a temporary workaround. Thanks.

**Comment by blinsay** [comment]
Is there documentation on disabling diagnostics in editors that aren't VSCode? I can't tell from this thread if `rust-analyzer.diagnostics.disabled` is a VSCode option or a rust-analyzer option.

Thanks for rust-analyzer, it's wonderful.

**Comment by lnicola** [comment]
It's a rust-analyzer option, but configuring it is client-dependant. If you're using coc.vim, you can use the same setting, otherwise it might be called diagnostics/enabled, without the program name.

If you figure it out and have the time, we'll probably accept client-specific documentation pull requests.

**Comment by matklad** [comment]
Some discussion here: https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/Where.20are.20third.20party.20crates.20in.20rustc-src.3F

**Comment by matklad** [comment]
https://github.com/rust-analyzer/rust-analyzer/pull/6294 papers over this issue

**Comment by jonas-schievink** [comment]
This still seems to happen, despite the workaround. Or does that only work if something else pulls in a `cfg_if` dependency?

**Comment by matklad** [comment]
Yeah, `cfg_if` should be in dep tree somewhere

**Comment by iamwwc** [comment]
Still showing this problem when I use `use std::os::unix::io::{AsRawFd, RawFd};` :(
Please fix it, think you :)

**Comment by iamwwc** [comment]
Mio seems works in WSL2

![image](https://user-images.githubusercontent.com/24750337/99547666-c0e52800-29f2-11eb-9083-dd416433dc3c.png)

But my project doesn't show docs and hint.

![image](https://user-images.githubusercontent.com/24750337/99547772-e4a86e00-29f2-11eb-9e86-ec1f77c2f2d4.png)

**Comment by talk2drys** [comment]
same issue. even though it works

![image](https://user-images.githubusercontent.com/32944188/100399282-63169880-300f-11eb-9a71-a4f2ae7ebe78.png)

**Comment by danfo** [comment]
In my projects I have unresolved imports reported on files deep in `~/.rustup/`. Sounds like it could be this known issue?

<details>

![Screen Shot 2020-12-07 at 9 09 17 am](https://user-images.githubusercontent.com/99922/101295894-08313c80-386c-11eb-9864-22eee8c2adbc.png)

</details>

**Comment by AXIM0S** [comment]
Same issue when using `std::os::unix::fs::symlink`.

**Comment by 2deth** [comment]
Resolved?! 😳

![image](https://user-images.githubusercontent.com/22878731/103147004-f5879480-4793-11eb-9c04-8fcde1a38a39.png)

```fish
❯ rust-analyzer --version
rust-analyzer 5e2935e
```

**Comment by bjorn3** [comment]
This was worked around in #6294.

**Comment by lnicola** [comment]
But I think we've had reports of it even after #6294 + one week.

**Comment by GrayJack** [comment]
I think it's because cfg_if crate must be in the dependency tree for it to work

**Comment by 2deth** [comment]
> This was worked around in #6294.

Hmm it didn’t work last week.. iirc

**Comment by 2deth** [comment]
"unresolved import"s are back.. ​😢​

I suspect that "resolved" happened after I put `%`s to `target-dir`..
And yes `cfg_if`.. `cargo build` has failed on that today.

But I cannot reproduce "resolved".
I use nightly and update everyday, and messed up many things recently.

Sorry for the wrong report.

**Comment by mvforell** [comment]
@jonas-schievink you wrote:
> Oh, it's because we don't load the sysroot correctly. We just add a bunch of hard-coded path dependencies to libstd without looking at its `Cargo.toml`.

I'm not familiar with how `rust_analyzer`'s works, but quickly skimming through the code I found [these lines](https://github.com/rust-analyzer/rust-analyzer/blob/86f947c1fb61c17140653f43e5f6e3e5c0ed057d/crates/project_model/src/sysroot.rs#L165-L190) where this hardcoding seems to reside. Without exactly knowing what I'm doing I added `cfg_if` to both strings, but sadly that does not seem to fix the issue.

Maybe you/someone could explain why it's not so simple, to make it easier to look into this issue?

**Comment by jonas-schievink** [comment]
`cfg_if` is a crates.io dependency, so it isn't available in the sysroot. Those lists only determine how the dependency graph within sysroot crates looks.

One possible way to solve this problem would be to vendor all external dependencies in the `rust-src` component. Another would be to run `cargo metadata` on the sysroot, which will download all crates.io dependencies (though it might require some file system trickery to avoid polluting rustup-managed directories).

**Comment by thomcc** [comment]
What would be needed to improve this in rust-analyzer? Is it possible or is it a fundamental limitation?

I'd be willing to try and work on it if possible, even if it's not exactly beginner-level. It is a ever-present pain point for me.

**Comment by flodiebold** [comment]
Maybe we could just vendor [the definition of cfg_if](https://github.com/alexcrichton/cfg-if/blob/349913689695130c2c63d1075e57b0917c98ee6e/src/lib.rs#L34-L100) ourselves in rust-analyzer, to make the workaround at least work in all cases.

**Comment by inflation** [comment]
I guess this is a great point why `std` should really have this macro in itself.f

**Comment by thomcc** [comment]
> Maybe we could just vendor [the definition of cfg_if](https://github.com/alexcrichton/cfg-if/blob/349913689695130c2c63d1075e57b0917c98ee6e/src/lib.rs#L34-L100) ourselves in rust-analyzer, to make the workaround at least work in all cases.

It doesn't actually work for me even if I explicitly add `cfg_if` into my dep tree. Like, 

The following errors
```rust
#[cfg(target_arch = "x86_64")]
use core::arch::x86_64; 
```
- regardless of std vs core.
- regardless of cfg_if as a dep or not. (I've even tried bringing cfg_if in normally vs `#[macro_use]`, ensuring I've used cfg_if before the code in question, etc).
- regardless of importing submodules, aliasing the import, etc.
- regardless of whether or not I `cfg` the import (although if the cfg doesnt apply there is no error, of course, but I can't use things).
- even if I put the `use` itself in a `cfg_if` rather than explictly putting it behind a `cfg`.

One thing that *doesn't* error, but I wouldn't phrase as rust-analyzer actually working (it has no semantic knowledge but also no red squiggles) is when I put that import inside a function. At function scope rust-analyzer seems to fail to resolve the module, but doesn't mark it as erroring, e.g.

```rust
// partially works — syntax highlighting 
// is slightly wrong inside the function
// and no semantic support is available,
// but at least it doesn't error
fn foo() {
	#[cfg(target_arch = "x86_64")]
	use core::arch::x86_64 as arch;
	#[cfg(target_arch = "x86")]
	use core::arch::x86 as arch;
}
```
(Example of "slightly wrong" syntax highlighting: https://user-images.githubusercontent.com/860665/107137781-b8e7a680-68c4-11eb-91d8-77ef140a2bc2.png — normally the parts after `x` wouldn't be white)

While this is way better than it giving an error, it's still unfortunate as r-a failing to understand the code seems to be (unsurprisingly) somewhat viral, and pretty soon a lot of your code isn't understood.

So yeah, the workaround with cfg_if either broke, never worked, never worked in this case, ...

---

Tangentially...

> I guess this is a great point why std should really have this macro in itself.

I used to agree but I think it's kinda a wrong way to structure stuff now in a lot of cases. Now I mostly just write the `#[cfg(...)]`, which works much better with r-a. (Well, maybe my opinion would be different if these options all worked with it though). 

Also, just purely aesthetically I think something more like:

```rust
cfg_match! {
	#[cfg(blah)] => { ... }
	#[cfg(foo)] => { ... }
}
```

Is a cleaner API (no repetition for "if"), and more natural expression support (although AFAIK none of the crates that implement this have expression support that isn't a little janky). Anyway, regardless of whether or not you agree, stabilizing it would have to go through this bikeshed-color-picking process...

Also note that `core::arch` is a dep of libcore (well... it's complicated) so libcore exposing `cfg_if!` might not help anyway, I'm not really sure.

**Comment by bjorn3** [comment]
`core::arch` not resolving is a different issue. This one is due to `#[path = "..."] mod arch;` not working when the path is not part of any directory assumed to be part of a crate in the crate graph based on the output of `cargo metadata`.

**Comment by thomcc** [comment]
> core::arch not resolving is a different issue. This one is due to #[path = "..."] mod arch; not working when the path is not part of any directory assumed to be part of a crate in the crate graph based on the output of cargo metadata.

Oh. Well. Is there a separate bug for that? How tractable is it to fix in r-a?

If it's very hard to fix in r-a I could do a patch to change stdarch itself to not use #[path] in the name of simplifying things for tooling probably, since usually the alternative there is just a slightly more complex `mod`/`use` setup.

**Comment by lnicola** [comment]
@jonas-schievink this doesn't seem to work for me even after https://github.com/rust-lang/rust/pull/81969/:

![image](https://user-images.githubusercontent.com/308347/109025330-4ef23e00-76c7-11eb-8f1e-ac7480d8f5e7.png)

![image](https://user-images.githubusercontent.com/308347/109025368-5580b580-76c7-11eb-843a-94d1de72087d.png)

**Comment by lnicola** [comment]
Oh, there's plenty more in `lib/rustlib/src/rust/library/std/src/sys/mod.rs`

**Comment by jonas-schievink** [comment]
Ugh, how annoying

**Comment by vberger** [comment]
Has something happened? I was successfully using the "add cfg-if in your dependency tree" method, but since today it no longer works, I again get unresolved import errors in vscode for `std::os::unix::*`.

Version 0.2.513 of the vscode rust-analyzer extension, and `rustc 1.52.0-nightly (acca81892 2021-03-13)`

**Comment by nico-abram** [comment]
I'm encountering this issue when doing `use std::os::windows::ffi::OsStrExt;` (With no conditional compilation attributes, it's a toy project that will only run in my machine). I don't have cfg_if in my dep tree

**Comment by CDirkx** [comment]
rust-lang/rust#84200 just got merged, which also removed more `cfg_if` in publicly exported items. I believe this should make everything in `std::os` usable, without needing to add cfg-if to your dependency tree. Can anyone confirm if this is the case?

**Comment by jonas-schievink** [comment]
Looks like this block still causes problems with `std::os::unix`: https://github.com/rust-lang/rust/blob/bacf770f2983a52f31e3537db5f0fe1ef2eaa874/library/std/src/sys/mod.rs#L27-L50

However, `std::os::linux` seems to work fine now :tada:

**Comment by noproto** [comment]
I'm no longer seeing any errors! (latest toolchain Rust 1.53, nightly rust-analyzer) 🥳🥳🥳

**Comment by Meister1593** [comment]
Works for me only with git rust-analyzer (pretty much current nightly) and nightly rust 1.55 (nightly-x86_64-unknown-linux-gnu, 1.53  stable didn't work)

**Comment by jonas-schievink** [comment]
Cool, looks like we can close this then. The more general problem on the r-a side is tracked in https://github.com/rust-analyzer/rust-analyzer/issues/7637.

**Comment by xanderio** [comment]
I'm still having this issue on my FreeBSD box. 
`rustc 1.54.0 (a178d0322 2021-07-26)`
`rust-analyzer bc084a6ee 2021-08-08 dev`

**Comment by jackloomen** [comment]
use rayon::prelude::*;

4:5   error   unresolved import ​rust-analyzer:

**Comment by ChayimFriedman2** [comment]
I think we can remove the hack by now, no? IIRC we now load the sysroot correctly.

**Comment by Veykril** [comment]
Yes we can remove this now

**Comment by Astlaan** [comment]
Don't know if related, but    

```rust
#[cfg(target_family = "unix")]
    let mut signal = tokio::signal::unix::signal(tokio::signal::unix::SignalKind::terminate())
        .expect("Failed to create signal handler");

#[cfg(windows)]
let mut signal = tokio::signal::windows::ctrl_close().expect("Failed to create signal handler");
```

I am developing on windows. Apparently I can't inspect `tokio::signal::unix` and its submodules. I also get no inline type hints for signal. How to do cross platform development?

**Comment by ChayimFriedman2** [comment]
@Astlaan This is not related. This issue is about cfgs in std not resolved correctly. Yours is because we don't provide IDE services for cfg'ed out code.

**Comment by Astlaan** [comment]
> [@Astlaan](https://github.com/Astlaan) This is not related. This issue is about cfgs in std not resolved correctly. Yours is because we don't provide IDE services for cfg'ed out code.

@ChayimFriedman2 
I see. Is this a planned feature? Or what is the typically strategy for developing code in windows, that runs both in windows and linux?

**Comment by ChayimFriedman2** [comment]
@Astlaan This is not something I am happy with (I think RustRover does offer IDE services for cfg'ed out code), but it's a big change and I can make no promises as to if/when this will get improved. Currently I'm just developing on the same OS, if this is a big chunk of platform-dependent code (you need to test it on the target OS anyway).
